### ISSUE TITLE
questionable write_literal warning: write!(f, "{}", "{integer}")

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6768]
Lint name: write_literal

I tried this code:

```rust
use std::fmt;

struct S {}

impl fmt::Display for S {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", "{integer}")
    }
}

fn main() {
    println!("{}", S {});
}
```

```
warning: literal with an empty format string
 --> src/main.rs:7:25
  |
7 |         write!(f, "{}", "{integer}")
  |                         ^^^^^^^^^^^
  |
  = note: `#[warn(clippy::write_literal)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#write_literal
  ```

In this code we use dummy format string `{}` to insert the string literal "{integer}" because if we had
`write!(f, "{ìnteger}")`, the `{integer}` would not be parsed as a raw literal(?):
````
error: there is no argument named `integer`
 --> src/main.rs:7:20
  |
7 |         write!(f, "{integer}")
  |                    ^^^^^^^^^
````
 
This is quite an interesting edge-case :)

I'm not sure what the best solution is here, if we can somehow provide a better suggestion or should just ignore literals that enclose anything in curly braces.

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (9b471a3 2021-02-19)`
- `rustc -Vv`:
  ```
rustc 1.52.0-nightly (9b471a3f5 2021-02-19)
binary: rustc
commit-hash: 9b471a3f5fe57e5c6e08acf665f2094422415a3d
commit-date: 2021-02-19
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 11.0.1

  ```

**Comment by hyd-dev** [comment]
Unless I misunderstood, I can (and prefer to) escape the curly braces instead ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8b6947fbaf67c35be4773a2eb6d567d2)):
```rust
use std::fmt;

struct S {}

impl fmt::Display for S {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{{integer}}")
    }
}

fn main() {
    println!("{}", S {});
}
```
So... please do not "ignore literals that enclose anything in curly braces", at least provide a way to opt-in to the warning.

**Comment by matthiaskrgr** [comment]
Ah, I forgot about the escaping with double curly braces!
Shall I close this again? :)

**Comment by camsteffen** [comment]
No, this is still a false positive. I think this is related to `pre_expansion_pass`.
